flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java

```diff
@@ -0,0 +1,192 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.api.java.tuple.Tuple2;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A {@code SharedStateRegistry} will be deployed in the 
+ * {@link org.apache.flink.runtime.checkpoint.CheckpointCoordinator} to 
+ * maintain the reference count of those state objects shared among different
+ * checkpoints. Each shared state object must be identified by a unique key. 
+ */
+public class SharedStateRegistry implements Serializable {
+
+	private static final long serialVersionUID = -8357254413007773970L;
+
+	/** All registered state objects */
+	private final Map<String, Tuple2<StateObject, Integer>> registeredStates = new HashMap<>();
+
+	/** All state objects that are not referenced any more */
+	private transient final List<StateObject> discardedStates = new ArrayList<>();
+
+	/**
+	 * Register the state in the registry
+	 *
+	 * @param key The key of the state to register
+	 * @param state The state to register
+	 */
+	public void register(String key, StateObject state) {
+		Tuple2<StateObject, Integer> stateAndRefCnt = registeredStates.get(key);
+
+		if (stateAndRefCnt == null) {
+			registeredStates.put(key, new Tuple2<>(state, 1));
+		} else {
+			if (!stateAndRefCnt.f0.equals(state)) {
+				throw new IllegalStateException("Cannot register a key with different states.");
+			}
+
+			stateAndRefCnt.f1++;
+		}
+	}
+
+	/**
+	 * Decrease the reference count of the state in the registry
+	 *
+	 * @param key The key of the state to unregister
+	 */
+	public void unregister(String key) {
+		Tuple2<StateObject, Integer> stateAndRefCnt = registeredStates.get(key);
+
+		if (stateAndRefCnt == null) {
+			throw new IllegalStateException("Cannot unregister an unexisted state.");
+		}
+
+		stateAndRefCnt.f1--;
+
+		// Remove the state from the registry when it's not referenced any more.
+		if (stateAndRefCnt.f1 == 0) {
+			registeredStates.remove(key);
+			discardedStates.add(stateAndRefCnt.f0);
+		}
+	}
+
+	/**
+	 * Register all the shared states in the given state handles.
+	 * 
+	 * @param stateHandles The state handles to register their shared states
+	 */
+	public void registerAll(Collection<? extends CompositeStateHandle> stateHandles) {
+		synchronized (this) {
```

Comment:
It is good practice to use a dedicated lock object, because someone outside may hold a lock on `this`.